src/de.hpi.swa.graal.squeak/src/de/hpi/swa/graal/squeak/nodes/ExecuteContextNode.java

```diff
@@ -58,11 +60,56 @@
     @Child private GetOrCreateContextNode getOrCreateContextNode;
 
     @Child private FrameStackInitializationNode frameInitializationNode;
-    @Child private HandlePrimitiveFailedNode handlePrimitiveFailedNode;
     @Child private MaterializeContextOnMethodExitNode materializeContextOnMethodExitNode;
 
     private SourceSection section;
 
+    private static final class ExecutePrimitiveContextNode extends ExecuteContextNode {
+
+        @Child private HandlePrimitiveFailedNode handlePrimitiveFailedNode;
+
+        ExecutePrimitiveContextNode(final CompiledCodeObject code, final boolean resume) {
+            super(code, resume);
+            assert !resume || code.isUnwindMarked() || code instanceof CompiledMethodObject && ((CompiledMethodObject) code).isExceptionHandlerMarked();
+        }
+
+        @Override
+        public Object executeFresh(final VirtualFrame frame) {
+            super.initialize(frame);
+            final CallPrimitiveNode callPrimitiveNode = (CallPrimitiveNode) fetchNextBytecodeNode(0);
+            if (callPrimitiveNode.primitiveNode != null) {
+                try {
+                    return callPrimitiveNode.primitiveNode.executePrimitive(frame);
+                } catch (final PrimitiveFailed e) {
+                    getHandlePrimitiveFailedNode().executeHandle(frame, e.getReasonCode());
+                    LOG.log(Level.FINE, () -> callPrimitiveNode.primitiveNode +
+                                    " (" + ArrayUtils.toJoinedString(", ", FrameAccess.getReceiverAndArguments(frame)) + ")");
+                    /* continue with fallback code. */
+                }
+            }
+            assert callPrimitiveNode.getSuccessorIndex() == CallPrimitiveNode.NUM_BYTECODES;
+            return super.executeFresh(frame);
```

Comment:
Subclassing is potentially dangerous in Truffle. `ExecuteContextNode#executeFresh` is no longer final, so the compiler might use an invoke here if you are unlucky.